File: bpf/aiml/cuda_tracer.c
// Copyright The Telegen Authors
// SPDX-License-Identifier: Apache-2.0
// Task: ML-014 - CUDA Kernel Tracer eBPF

#include "../bpfcore/vmlinux.h"
#include "../bpfcore/bpf_helpers.h"
#include "../bpfcore/bpf_tracing.h"
#include "../bpfcore/bpf_core_read.h"
#include "../common/common.h"

// CUDA event types
#define CUDA_EVENT_KERNEL_LAUNCH     0
#define CUDA_EVENT_KERNEL_COMPLETE   1
#define CUDA_EVENT_MEMCPY_START      2
#define CUDA_EVENT_MEMCPY_COMPLETE   3
#define CUDA_EVENT_MALLOC            4
#define CUDA_EVENT_FREE              5
#define CUDA_EVENT_SYNC              6
#define CUDA_EVENT_STREAM_CREATE     7
#define CUDA_EVENT_STREAM_DESTROY    8

// Memory copy direction
#define CUDA_MEMCPY_HOST_TO_HOST     0
#define CUDA_MEMCPY_HOST_TO_DEVICE   1
#define CUDA_MEMCPY_DEVICE_TO_HOST   2
#define CUDA_MEMCPY_DEVICE_TO_DEVICE 3
#define CUDA_MEMCPY_DEFAULT          4

#define CUDA_MAX_KERNEL_NAME 128

// CUDA kernel event structure
struct cuda_event {
    u64 timestamp_ns;        // Event timestamp
    u64 duration_ns;         // Duration (for completion events)
    u32 pid;                 // Process ID
    u32 tid;                 // Thread ID
    u32 event_type;          // Event type
    u32 gpu_id;              // GPU device ID
    u64 stream_id;           // CUDA stream ID
    
    // Kernel launch info
    u32 grid_dim_x;
    u32 grid_dim_y;
    u32 grid_dim_z;
    u32 block_dim_x;
    u32 block_dim_y;
    u32 block_dim_z;
    u32 shared_mem_bytes;
    u8 kernel_name[CUDA_MAX_KERNEL_NAME];
    
    // Memory operation info
    u64 src_ptr;
    u64 dst_ptr;
    u64 bytes;
    u32 memcpy_kind;
    
    // Memory allocation info
    u64 alloc_ptr;
    u64 alloc_size;
    
    // Error info
    u32 cuda_error;
};

// Active kernel tracking
struct cuda_kernel_info {
    u64 start_time;
    u32 pid;
    u32 tid;
    u32 gpu_id;
    u64 stream_id;
    u32 grid_dim_x;
    u32 grid_dim_y;
    u32 grid_dim_z;
    u32 block_dim_x;
    u32 block_dim_y;
    u32 block_dim_z;
    u32 shared_mem_bytes;
    u8 kernel_name[CUDA_MAX_KERNEL_NAME];
};

// Active memory operation tracking
struct cuda_memop_info {
    u64 start_time;
    u32 pid;
    u32 tid;
    u64 src_ptr;
    u64 dst_ptr;
    u64 bytes;
    u32 memcpy_kind;
    u64 stream_id;
};

// Map to track active CUDA kernels
struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 10240);
    __type(key, u64);  // kernel correlation ID
    __type(value, struct cuda_kernel_info);
} cuda_active_kernels SEC(".maps");

// Map to track active memory operations
struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 10240);
    __type(key, u64);  // operation correlation ID
    __type(value, struct cuda_memop_info);
} cuda_active_memops SEC(".maps");

// Ring buffer for CUDA events
struct {
    __uint(type, BPF_MAP_TYPE_RINGBUF);
    __uint(max_entries, 1 << 21);  // 2MB buffer
} cuda_events SEC(".maps");

// Per-process GPU memory tracking
struct cuda_mem_stats {
    u64 total_allocated;
    u64 total_freed;
    u64 peak_usage;
    u64 current_usage;
    u64 alloc_count;
    u64 free_count;
};

struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 4096);
    __type(key, u32);  // PID
    __type(value, struct cuda_mem_stats);
} cuda_mem_per_process SEC(".maps");

// Kernel name cache
struct {
    __uint(type, BPF_MAP_TYPE_LRU_HASH);
    __uint(max_entries, 4096);
    __type(key, u64);  // Function pointer
    __type(value, u8[CUDA_MAX_KERNEL_NAME]);
} cuda_kernel_names SEC(".maps");

// Submit a CUDA event to the ring buffer
static __always_inline int submit_cuda_event(struct cuda_event *event) {
    struct cuda_event *e = bpf_ringbuf_reserve(&cuda_events, sizeof(*event), 0);
    if (!e) {
        return -1;
    }
    
    __builtin_memcpy(e, event, sizeof(*event));
    bpf_ringbuf_submit(e, 0);
    return 0;
}

// Track cudaLaunchKernel calls
SEC("uprobe/cudaLaunchKernel")
int BPF_UPROBE(cudaLaunchKernel, 
               void *func,        // Kernel function pointer
               u32 grid_x, u32 grid_y, u32 grid_z,
               u32 block_x, u32 block_y, u32 block_z,
               void **args,
               u64 shared_mem,
               void *stream) {
    
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tid = pid_tgid;
    u64 now = bpf_ktime_get_ns();
    
    // Create kernel tracking entry
    struct cuda_kernel_info info = {};
    info.start_time = now;
    info.pid = pid;
    info.tid = tid;
    info.stream_id = (u64)stream;
    info.grid_dim_x = grid_x;
    info.grid_dim_y = grid_y;
    info.grid_dim_z = grid_z;
    info.block_dim_x = block_x;
    info.block_dim_y = block_y;
    info.block_dim_z = block_z;
    info.shared_mem_bytes = (u32)shared_mem;
    
    // Look up kernel name from cache
    u64 func_ptr = (u64)func;
    u8 *name = bpf_map_lookup_elem(&cuda_kernel_names, &func_ptr);
    if (name) {
        __builtin_memcpy(info.kernel_name, name, CUDA_MAX_KERNEL_NAME);
    }
    
    // Store in active kernels map
    u64 kernel_id = (pid_tgid << 16) | (now & 0xFFFF);
    bpf_map_update_elem(&cuda_active_kernels, &kernel_id, &info, BPF_ANY);
    
    // Submit launch event
    struct cuda_event event = {};
    event.timestamp_ns = now;
    event.pid = pid;
    event.tid = tid;
    event.event_type = CUDA_EVENT_KERNEL_LAUNCH;
    event.stream_id = (u64)stream;
    event.grid_dim_x = grid_x;
    event.grid_dim_y = grid_y;
    event.grid_dim_z = grid_z;
    event.block_dim_x = block_x;
    event.block_dim_y = block_y;
    event.block_dim_z = block_z;
    event.shared_mem_bytes = (u32)shared_mem;
    __builtin_memcpy(event.kernel_name, info.kernel_name, CUDA_MAX_KERNEL_NAME);
    
    submit_cuda_event(&event);
    
    return 0;
}

// Track cudaMemcpy calls
SEC("uprobe/cudaMemcpy")
int BPF_UPROBE(cudaMemcpy, void *dst, void *src, u64 count, u32 kind) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tid = pid_tgid;
    u64 now = bpf_ktime_get_ns();
    
    // Create memop tracking entry
    struct cuda_memop_info info = {};
    info.start_time = now;
    info.pid = pid;
    info.tid = tid;
    info.src_ptr = (u64)src;
    info.dst_ptr = (u64)dst;
    info.bytes = count;
    info.memcpy_kind = kind;
    
    u64 memop_id = (pid_tgid << 16) | (now & 0xFFFF);
    bpf_map_update_elem(&cuda_active_memops, &memop_id, &info, BPF_ANY);
    
    // Submit start event
    struct cuda_event event = {};
    event.timestamp_ns = now;
    event.pid = pid;
    event.tid = tid;
    event.event_type = CUDA_EVENT_MEMCPY_START;
    event.src_ptr = (u64)src;
    event.dst_ptr = (u64)dst;
    event.bytes = count;
    event.memcpy_kind = kind;
    
    submit_cuda_event(&event);
    
    return 0;
}

// Track cudaMemcpyAsync calls
SEC("uprobe/cudaMemcpyAsync")
int BPF_UPROBE(cudaMemcpyAsync, void *dst, void *src, u64 count, u32 kind, void *stream) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tid = pid_tgid;
    u64 now = bpf_ktime_get_ns();
    
    struct cuda_memop_info info = {};
    info.start_time = now;
    info.pid = pid;
    info.tid = tid;
    info.src_ptr = (u64)src;
    info.dst_ptr = (u64)dst;
    info.bytes = count;
    info.memcpy_kind = kind;
    info.stream_id = (u64)stream;
    
    u64 memop_id = (pid_tgid << 16) | (now & 0xFFFF);
    bpf_map_update_elem(&cuda_active_memops, &memop_id, &info, BPF_ANY);
    
    struct cuda_event event = {};
    event.timestamp_ns = now;
    event.pid = pid;
    event.tid = tid;
    event.event_type = CUDA_EVENT_MEMCPY_START;
    event.stream_id = (u64)stream;
    event.src_ptr = (u64)src;
    event.dst_ptr = (u64)dst;
    event.bytes = count;
    event.memcpy_kind = kind;
    
    submit_cuda_event(&event);
    
    return 0;
}

// Track cudaMalloc calls
SEC("uprobe/cudaMalloc")
int BPF_UPROBE(cudaMalloc, void **devPtr, u64 size) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tid = pid_tgid;
    u64 now = bpf_ktime_get_ns();
    
    struct cuda_event event = {};
    event.timestamp_ns = now;
    event.pid = pid;
    event.tid = tid;
    event.event_type = CUDA_EVENT_MALLOC;
    event.alloc_size = size;
    
    submit_cuda_event(&event);
    
    // Update memory statistics
    struct cuda_mem_stats *stats = bpf_map_lookup_elem(&cuda_mem_per_process, &pid);
    if (stats) {
        __sync_fetch_and_add(&stats->total_allocated, size);
        __sync_fetch_and_add(&stats->current_usage, size);
        __sync_fetch_and_add(&stats->alloc_count, 1);
        
        // Update peak if needed
        if (stats->current_usage > stats->peak_usage) {
            stats->peak_usage = stats->current_usage;
        }
    } else {
        struct cuda_mem_stats new_stats = {};
        new_stats.total_allocated = size;
        new_stats.current_usage = size;
        new_stats.peak_usage = size;
        new_stats.alloc_count = 1;
        bpf_map_update_elem(&cuda_mem_per_process, &pid, &new_stats, BPF_ANY);
    }
    
    return 0;
}

// Track cudaFree calls
SEC("uprobe/cudaFree")
int BPF_UPROBE(cudaFree, void *devPtr) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tid = pid_tgid;
    u64 now = bpf_ktime_get_ns();
    
    struct cuda_event event = {};
    event.timestamp_ns = now;
    event.pid = pid;
    event.tid = tid;
    event.event_type = CUDA_EVENT_FREE;
    event.alloc_ptr = (u64)devPtr;
    
    submit_cuda_event(&event);
    
    // Update memory statistics
    struct cuda_mem_stats *stats = bpf_map_lookup_elem(&cuda_mem_per_process, &pid);
    if (stats) {
        __sync_fetch_and_add(&stats->free_count, 1);
        // Note: We don't know the size being freed without tracking allocations
    }
    
    return 0;
}

// Track cudaDeviceSynchronize calls
SEC("uprobe/cudaDeviceSynchronize")
int BPF_UPROBE(cudaDeviceSynchronize) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tid = pid_tgid;
    u64 now = bpf_ktime_get_ns();
    
    struct cuda_event event = {};
    event.timestamp_ns = now;
    event.pid = pid;
    event.tid = tid;
    event.event_type = CUDA_EVENT_SYNC;
    
    submit_cuda_event(&event);
    
    return 0;
}

// Track cudaStreamCreate calls
SEC("uprobe/cudaStreamCreate")
int BPF_UPROBE(cudaStreamCreate, void **pStream) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tid = pid_tgid;
    u64 now = bpf_ktime_get_ns();
    
    struct cuda_event event = {};
    event.timestamp_ns = now;
    event.pid = pid;
    event.tid = tid;
    event.event_type = CUDA_EVENT_STREAM_CREATE;
    
    submit_cuda_event(&event);
    
    return 0;
}

// Track cudaStreamDestroy calls
SEC("uprobe/cudaStreamDestroy")
int BPF_UPROBE(cudaStreamDestroy, void *stream) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tid = pid_tgid;
    u64 now = bpf_ktime_get_ns();
    
    struct cuda_event event = {};
    event.timestamp_ns = now;
    event.pid = pid;
    event.tid = tid;
    event.event_type = CUDA_EVENT_STREAM_DESTROY;
    event.stream_id = (u64)stream;
    
    submit_cuda_event(&event);
    
    return 0;
}

// Return probe for cudaLaunchKernel to capture completion
SEC("uretprobe/cudaLaunchKernel")
int BPF_URETPROBE(cudaLaunchKernel_ret, int ret) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tid = pid_tgid;
    u64 now = bpf_ktime_get_ns();
    
    // Note: This captures the return from the launch, not the actual kernel completion
    // True kernel completion tracking requires CUPTI or synchronization tracking
    
    if (ret != 0) {
        struct cuda_event event = {};
        event.timestamp_ns = now;
        event.pid = pid;
        event.tid = tid;
        event.event_type = CUDA_EVENT_KERNEL_COMPLETE;
        event.cuda_error = ret;
        
        submit_cuda_event(&event);
    }
    
    return 0;
}

// Return probe for cudaMemcpy to capture completion
SEC("uretprobe/cudaMemcpy")
int BPF_URETPROBE(cudaMemcpy_ret, int ret) {
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tid = pid_tgid;
    u64 now = bpf_ktime_get_ns();
    
    struct cuda_event event = {};
    event.timestamp_ns = now;
    event.pid = pid;
    event.tid = tid;
    event.event_type = CUDA_EVENT_MEMCPY_COMPLETE;
    event.cuda_error = ret;
    
    submit_cuda_event(&event);
    
    return 0;
}

char LICENSE[] SEC("license") = "Dual BSD/GPL";

File: internal/tracers/cudatracer/cudatracer.go
// Copyright The Telegen Authors
// SPDX-License-Identifier: Apache-2.0

//go:build linux

// Package cudatracer provides eBPF-based CUDA kernel and memory operation tracing.
// Task: ML-014 - CUDA Kernel eBPF Tracer Userspace Loader
package cudatracer // import "github.com/platformbuilds/telegen/internal/tracers/cudatracer"

import (
	"context"
	"errors"
	"io"
	"log/slog"
	"sync"

	"github.com/cilium/ebpf"

	"github.com/platformbuilds/telegen/internal/appolly/app/request"
	"github.com/platformbuilds/telegen/internal/appolly/app/svc"
	"github.com/platformbuilds/telegen/internal/discover/exec"
	ebpfcommon "github.com/platformbuilds/telegen/internal/ebpf/common"
	"github.com/platformbuilds/telegen/internal/goexec"
	"github.com/platformbuilds/telegen/internal/obi"
	config "github.com/platformbuilds/telegen/internal/obiconfig"
	"github.com/platformbuilds/telegen/internal/ringbuf"
	"github.com/platformbuilds/telegen/pkg/export/imetrics"
	"github.com/platformbuilds/telegen/pkg/pipe/msg"
)

//go:generate $BPF2GO -cc $BPF_CLANG -cflags $BPF_CFLAGS -type cuda_event_t -type cuda_mem_stats_t -target amd64,arm64 Bpf ../../../bpf/aiml/cuda_tracer.c -- -I../../../bpf

// CUDA event types - must match constants in cuda_tracer.h
const (
	CUDAEventKernelLaunch   = 0
	CUDAEventKernelComplete = 1
	CUDAEventMemcpyStart    = 2
	CUDAEventMemcpyComplete = 3
	CUDAEventMalloc         = 4
	CUDAEventFree           = 5
	CUDAEventSync           = 6
	CUDAEventStreamCreate   = 7
	CUDAEventStreamDestroy  = 8
)

// Memory copy direction
const (
	CUDAMemcpyHostToHost     = 0
	CUDAMemcpyHostToDevice   = 1
	CUDAMemcpyDeviceToHost   = 2
	CUDAMemcpyDeviceToDevice = 3
	CUDAMemcpyDefault        = 4
)

// CUDAEvent wraps the BPF CUDA event type
type CUDAEvent BpfCudaEventT

// CUDAMemStats wraps the BPF memory statistics type
type CUDAMemStats BpfCudaMemStatsT

// Tracer implements eBPF-based CUDA kernel and memory operation tracing
type Tracer struct {
	pidsFilter       ebpfcommon.ServiceFilter
	cfg              *obi.Config
	metrics          imetrics.Reporter
	bpfObjects       BpfObjects
	closers          []io.Closer
	log              *slog.Logger
	instrumentedLibs ebpfcommon.InstrumentedLibsT
	libsMux          sync.Mutex
}

// New creates a new CUDA tracer
func New(pidFilter ebpfcommon.ServiceFilter, cfg *obi.Config, metrics imetrics.Reporter) *Tracer {
	log := slog.With("component", "cudatracer.Tracer")

	return &Tracer{
		log:              log,
		cfg:              cfg,
		metrics:          metrics,
		pidsFilter:       pidFilter,
		instrumentedLibs: make(ebpfcommon.InstrumentedLibsT),
		libsMux:          sync.Mutex{},
	}
}

// AllowPID adds a PID to the allowed list for tracing
func (p *Tracer) AllowPID(pid, ns uint32, svc *svc.Attrs) {
	p.pidsFilter.AllowPID(pid, ns, svc, ebpfcommon.PIDTypeKProbes)
}

// BlockPID removes a PID from the allowed list
func (p *Tracer) BlockPID(pid, ns uint32) {
	p.pidsFilter.BlockPID(pid, ns)
}

// Load returns the embedded BPF collection spec
func (p *Tracer) Load() (*ebpf.CollectionSpec, error) {
	return LoadBpf()
}

// Constants returns the constants to be injected into the BPF program
func (p *Tracer) Constants() map[string]any {
	m := make(map[string]any, 1)

	if p.cfg.Discovery.BPFPidFilterOff {
		m["filter_pids"] = int32(0)
	} else {
		m["filter_pids"] = int32(1)
	}

	return m
}

// RegisterOffsets is called when offsets are available for a binary
func (p *Tracer) RegisterOffsets(fileInfo *exec.FileInfo, _ *goexec.Offsets) {
	p.ProcessBinary(fileInfo)
}

// ProcessBinary processes a binary for CUDA instrumentation
func (p *Tracer) ProcessBinary(fileInfo *exec.FileInfo) {
	if fileInfo == nil || fileInfo.ELF == nil {
		p.log.Error("Empty fileinfo for CUDA tracer")
	}
}

// BpfObjects returns the BPF objects
func (p *Tracer) BpfObjects() any {
	return &p.bpfObjects
}

// AddCloser adds a closer to be called on cleanup
func (p *Tracer) AddCloser(c ...io.Closer) {
	p.closers = append(p.closers, c...)
}

// GoProbes returns Go probes - not used for CUDA tracer
func (p *Tracer) GoProbes() map[string][]*ebpfcommon.ProbeDesc {
	return nil
}

// KProbes returns kernel probes - not used for CUDA tracer
func (p *Tracer) KProbes() map[string]ebpfcommon.ProbeDesc {
	return nil
}

// Tracepoints returns tracepoints - not used for CUDA tracer
func (p *Tracer) Tracepoints() map[string]ebpfcommon.ProbeDesc {
	return nil
}

// UProbes returns the userspace probes for CUDA library functions
func (p *Tracer) UProbes() map[string]map[string][]*ebpfcommon.ProbeDesc {
	return map[string]map[string][]*ebpfcommon.ProbeDesc{
		"libcudart.so": {
			"cudaLaunchKernel": {{
				Start: p.bpfObjects.CudaLaunchKernel,
				End:   p.bpfObjects.CudaLaunchKernelRet,
			}},
			"cudaMemcpy": {{
				Start: p.bpfObjects.CudaMemcpy,
				End:   p.bpfObjects.CudaMemcpyRet,
			}},
			"cudaMemcpyAsync": {{
				Start: p.bpfObjects.CudaMemcpyAsync,
			}},
			"cudaMalloc": {{
				Start: p.bpfObjects.CudaMalloc,
			}},
			"cudaFree": {{
				Start: p.bpfObjects.CudaFree,
			}},
			"cudaDeviceSynchronize": {{
				Start: p.bpfObjects.CudaDeviceSynchronize,
			}},
			"cudaStreamCreate": {{
				Start: p.bpfObjects.CudaStreamCreate,
			}},
			"cudaStreamDestroy": {{
				Start: p.bpfObjects.CudaStreamDestroy,
			}},
		},
	}
}

// SetupTailCalls sets up tail calls - not used for CUDA tracer
func (p *Tracer) SetupTailCalls() {}

// SocketFilters returns socket filters - not used for CUDA tracer
func (p *Tracer) SocketFilters() []*ebpf.Program {
	return nil
}

// SockMsgs returns sock msg programs - not used for CUDA tracer
func (p *Tracer) SockMsgs() []ebpfcommon.SockMsg { return nil }

// SockOps returns sock ops programs - not used for CUDA tracer
func (p *Tracer) SockOps() []ebpfcommon.SockOps { return nil }

// Iters returns iterator programs - not used for CUDA tracer
func (p *Tracer) Iters() []*ebpfcommon.Iter { return nil }

// RecordInstrumentedLib records an instrumented library
func (p *Tracer) RecordInstrumentedLib(id uint64, closers []io.Closer) {
	p.libsMux.Lock()
	defer p.libsMux.Unlock()

	module := p.instrumentedLibs.AddRef(id)

	if len(closers) > 0 {
		module.Closers = append(module.Closers, closers...)
	}

	p.log.Debug("Recorded instrumented Lib", "ino", id, "module", module)
}

// AddInstrumentedLibRef adds a reference to an instrumented library
func (p *Tracer) AddInstrumentedLibRef(id uint64) {
	p.RecordInstrumentedLib(id, nil)
}

// UnlinkInstrumentedLib removes a reference to an instrumented library
func (p *Tracer) UnlinkInstrumentedLib(id uint64) {
	p.libsMux.Lock()
	defer p.libsMux.Unlock()

	module, err := p.instrumentedLibs.RemoveRef(id)

	p.log.Debug("Unlinking instrumented lib - before state", "ino", id, "module", module)

	if err != nil {
		p.log.Debug("Error unlinking instrumented lib", "ino", id, "error", err)
	}
}

// AlreadyInstrumentedLib checks if a library is already instrumented
func (p *Tracer) AlreadyInstrumentedLib(id uint64) bool {
	p.libsMux.Lock()
	defer p.libsMux.Unlock()

	module := p.instrumentedLibs.Find(id)

	p.log.Debug("checking already instrumented Lib", "ino", id, "module", module)
	return module != nil
}

// Run starts the tracer and forwards events to the output channel
func (p *Tracer) Run(ctx context.Context, ebpfEventContext *ebpfcommon.EBPFEventContext, eventsChan *msg.Queue[[]request.Span]) {
	ebpfcommon.ForwardRingbuf(
		&p.cfg.EBPF,
		p.bpfObjects.CudaEvents,
		ebpfEventContext.CommonPIDsFilter,
		p.processCUDAEvent,
		p.log,
		p.metrics,
		eventsChan,
		append(p.closers, &p.bpfObjects)...,
	)(ctx, eventsChan)
}

// processCUDAEvent parses a CUDA event from the ring buffer and converts it to a span
func (p *Tracer) processCUDAEvent(_ *ebpfcommon.EBPFParseContext, _ *config.EBPFTracer, record *ringbuf.Record, _ ebpfcommon.ServiceFilter) (request.Span, bool, error) {
	if len(record.RawSample) == 0 {
		return request.Span{}, true, errors.New("invalid ringbuffer record size")
	}

	event, err := ebpfcommon.ReinterpretCast[CUDAEvent](record.RawSample)
	if err != nil {
		return request.Span{}, true, err
	}

	span := request.Span{
		Pid: request.PidInfo{
			HostPID: event.Pid,
			UserPID: event.Pid,
		},
	}

	switch event.EventType {
	case CUDAEventKernelLaunch:
		span.Type = request.EventTypeGPUKernelLaunch
		span.Method = string(event.KernelName[:nullTerminated(event.KernelName[:])])
		// Grid size as content length (total threads = gridX * gridY * gridZ * blockX * blockY * blockZ)
		gridSize := int64(event.GridDimX) * int64(event.GridDimY) * int64(event.GridDimZ)
		blockSize := int64(event.BlockDimX) * int64(event.BlockDimY) * int64(event.BlockDimZ)
		span.ContentLength = gridSize * blockSize
		p.log.Debug("CUDA Kernel Launch", "kernel", span.Method, "threads", span.ContentLength)

	case CUDAEventMemcpyStart, CUDAEventMemcpyComplete:
		span.Type = request.EventTypeGPUMemcpy
		span.ContentLength = int64(event.Bytes)
		span.SubType = int(event.MemcpyKind)
		if event.EventType == CUDAEventMemcpyComplete {
			// Duration is calculated from End - Start in the span processing
			span.End = span.Start + int64(event.DurationNs)
		}
		p.log.Debug("CUDA Memcpy", "bytes", event.Bytes, "kind", event.MemcpyKind)

	case CUDAEventMalloc:
		span.Type = request.EventTypeGPUMalloc
		span.ContentLength = int64(event.AllocSize)
		p.log.Debug("CUDA Malloc", "size", event.AllocSize)

	case CUDAEventFree:
		span.Type = request.EventTypeGPUMalloc // Reuse same type for free
		span.ContentLength = 0                 // Free doesn't have size info from kernel
		span.SubType = 1                       // Mark as free operation
		p.log.Debug("CUDA Free", "ptr", event.AllocPtr)

	case CUDAEventSync:
		span.Type = request.EventTypeGPUKernelLaunch // Reuse for sync
		span.Method = "cudaDeviceSynchronize"
		p.log.Debug("CUDA Device Synchronize")

	case CUDAEventStreamCreate:
		span.Type = request.EventTypeGPUKernelLaunch
		span.Method = "cudaStreamCreate"
		p.log.Debug("CUDA Stream Create")

	case CUDAEventStreamDestroy:
		span.Type = request.EventTypeGPUKernelLaunch
		span.Method = "cudaStreamDestroy"
		p.log.Debug("CUDA Stream Destroy")

	default:
		p.log.Error("Unknown CUDA event type", "type", event.EventType)
		return request.Span{}, true, nil
	}

	return span, false, nil
}

// nullTerminated finds the index of the first null byte
func nullTerminated(b []byte) int {
	for i, c := range b {
		if c == 0 {
			return i
		}
	}
	return len(b)
}

// Required returns whether this tracer is required
func (p *Tracer) Required() bool {
	return false
}
